internal/controller/supervisorconfig/upstreamwatcher/ldap_upstream_watcher.go

```diff
@@ -0,0 +1,367 @@
+		withInformer(
+			ldapIdentityProviderInformer,
+			pinnipedcontroller.MatchAnythingFilter(pinnipedcontroller.SingletonQueue()),
+			controllerlib.InformerOption{},
+		),
+		withInformer(
+			secretInformer,
+			pinnipedcontroller.MatchAnySecretOfTypeFilter(ldapBindAccountSecretType, pinnipedcontroller.SingletonQueue()),
+			controllerlib.InformerOption{},
+		),
+	)
+}
+
+// Sync implements controllerlib.Syncer.
+func (c *ldapWatcherController) Sync(ctx controllerlib.Context) error {
+	actualUpstreams, err := c.ldapIdentityProviderInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list LDAPIdentityProviders: %w", err)
+	}
+
+	requeue := false
+	validatedUpstreams := make([]provider.UpstreamLDAPIdentityProviderI, 0, len(actualUpstreams))
+	for _, upstream := range actualUpstreams {
+		valid := c.validateUpstream(ctx.Context, upstream)
+		if valid == nil {
+			requeue = true
+		} else {
+			validatedUpstreams = append(validatedUpstreams, valid)
+		}
+	}
+	c.cache.SetLDAPIdentityProviders(validatedUpstreams)
+	if requeue {
+		return controllerlib.ErrSyntheticRequeue
+	}
+	return nil
+}
+
+func (c *ldapWatcherController) validateUpstream(ctx context.Context, upstream *v1alpha1.LDAPIdentityProvider) provider.UpstreamLDAPIdentityProviderI {
+	spec := upstream.Spec
+
+	config := &upstreamldap.ProviderConfig{
+		Name: upstream.Name,
+		Host: spec.Host,
+		UserSearch: upstreamldap.UserSearchConfig{
+			Base:              spec.UserSearch.Base,
+			Filter:            spec.UserSearch.Filter,
+			UsernameAttribute: spec.UserSearch.Attributes.Username,
+			UIDAttribute:      spec.UserSearch.Attributes.UniqueID,
+		},
+		Dialer: c.ldapDialer,
+	}
+
+	conditions := []*v1alpha1.Condition{}
+	secretValidCondition, currentSecretVersion := c.validateSecret(upstream, config)
+	tlsValidCondition := c.validateTLSConfig(upstream, config)
+	conditions = append(conditions, secretValidCondition, tlsValidCondition)
+
+	// No point in trying to connect to the server if the config was already determined to be invalid.
+	if secretValidCondition.Status == v1alpha1.ConditionTrue && tlsValidCondition.Status == v1alpha1.ConditionTrue {
+		finishedConfigCondition := c.validateFinishedConfig(ctx, upstream, config, currentSecretVersion)
+		// nil when there is no need to update this condition.
+		if finishedConfigCondition != nil {
+			conditions = append(conditions, finishedConfigCondition)
+		}
+	}
+
+	hadErrorCondition := c.updateStatus(ctx, upstream, conditions)
+	if hadErrorCondition {
+		return nil
+	}
+
+	return upstreamldap.New(*config)
+}
+
+func (c *ldapWatcherController) validateTLSConfig(upstream *v1alpha1.LDAPIdentityProvider, config *upstreamldap.ProviderConfig) *v1alpha1.Condition {
+	tlsSpec := upstream.Spec.TLS
+	if tlsSpec == nil {
+		return c.validTLSCondition(noTLSConfigurationMessage)
+	}
+	if len(tlsSpec.CertificateAuthorityData) == 0 {
+		return c.validTLSCondition(loadedTLSConfigurationMessage)
+	}
```

Comment:
This seems like it should error?

Reply:
`tlsSpec.CertificateAuthorityData` is optional. When it is missing we are careful to use the OS CA bundle.

Reply:
But why do we want this to work:

```yaml
...
host: example.com
tls: {}
...
```

i.e. this is more clear:

```yaml
...
host: example.com
...
```

Reply:
I can buy this reasoning: https://github.com/vmware-tanzu/pinniped/pull/588#discussion_r622415626 so I think we can consider this resolved.  You may want to use that `CABundle` func for consistency.